http2: fix graceful session close

pandeykushagra51 opened the pull request:
Fix issue where session.close() prematurely destroys the session when response.end() was called with an empty payload while active http2 streams still existed. This change ensures that sessions are closed gracefully only after all http2 streams complete and clients properly receive the GOAWAY frame as per the HTTP/2 spec.

Refs: https://nodejs.org/api/http2.html\#http2sessionclosecallback
Fixes: #57809 

## Update:
Please refer detailed explanation below to know the intention behind updating `test/parallel/test-http2-client-rststream-before-connect.js` test.
https://github.com/nodejs/node/pull/57808#issuecomment-2799023031

pandeykushagra51 commented:
###Problem:
When a client calls response.end() with an empty payload, Node.js sends only header frames to nghttp2, and the JavaScript layer assumes these frames will be successfully transmitted. If session.close() is called immediately afterward, the Http2Session class doesn't recognize any active streams (as it only tracks data frames and push promises as active streams), and begins terminating the underlying socket. However, when the C++ layer subsequently attempts to transmit the queued header frames, it fails because the socket has already been closed, resulting in transmission errors and clients never receiving essential GOAWAY frames and response header.

###Solution:
The solution ensures proper graceful session closure by delaying socket termination until all data in the nghttp2 queue/buffer is fully processed. This is implemented by:
Adding checks for nghttp2_session_want_write and nghttp2_session_want_read before closing the underlying socket
If nghttp2 has pending data to transmit, postponing the socket closure
Implementing a callback mechanism to notify the JavaScript layer when all writes have completed
Only proceeding with socket closure when all transmission conditions are satisfied
This approach guarantees that all frames, including header-only responses and GOAWAY frames, are properly delivered to the client before connection termination, ensuring robust HTTP/2 protocol compliance.

pandeykushagra51 commented:
I will be very happy to update the current solution in case of any scope of improvement and would love to take input to optimise this further.

pandeykushagra51 commented:
### Optimization Proposal for the current fix

## Current Issue:
The JavaScript layer is being called after each write operation
This creates overhead with repetitive calls to onstreamaftrwrite_string()

## Proposed Solution
Flag-Based Approach:
Set a flag in the C++ layer when JavaScript initiates a graceful closure
C++ layer checks this flag during write operations.

## Responsibility Shift:
When the flag is set, make the C++ layer responsible for notifying the JavaScript layer
Notification happens either:
After each write operation, OR
After nghttp2 completes all read and write operations

## Benefits:
Reduces repetitive calls to onstreamaftrwrite_string()
Improves performance by minimizing JS-C++ boundary crossings

## Trade-off:
Implementation will be more complex
Solution will be longer but more efficient

pandeykushagra51 commented on lib/internal/http2/core.js:
```diff
@@ -3300,7 +3303,7 @@ function socketOnClose() {
     state.streams.forEach((stream) => stream.close(NGHTTP2_CANCEL));
     state.pendingStreams.forEach((stream) => stream.close(NGHTTP2_CANCEL));
     session.close();
-    session[kMaybeDestroy](err);
+    closeSession(session, NGHTTP2_NO_ERROR, err);
```
This change is done as improvement. This is because when underlying socket is closed there is no need for looking at graceful closure of session by calling [kMaybeDestroy] instead we can immediately call closeSession which will handle all cleanup operation.

pandeykushagra51 commented:
One of the simplest approach is to delay socket destruction to next iteration of event loop i.e call inside setImmediate but I don't think this will address underlying problem

mcollina commented:
Some tests/linting are failing.

I think the proposed optimization would be useful.

pandeykushagra51 commented:
> Some tests/linting are failing.
> 
> I think the proposed optimization would be useful.

Fine, then I will be implementing more optimised solution as stated above.
Also regarding failing test case, I am running the test on MacOS with below system
```Darwin Mac.lan 24.3.0 Darwin Kernel Version 24.3.0: Thu Jan  2 20:24:24 PST 2025; root:xnu-11215.81.4~3/RELEASE_ARM64_T6030 arm64```
And all tests are passed every time, so I am a bit confused why it is failing on CI.
I will check for rest of the OS and submit report here.

pandeykushagra51 commented:
> Some tests/linting are failing.
> 
> I think the proposed optimization would be useful.

hey @mcollina  I have implemented following strategy for optimisation:
1. Introduced a graceful session closed flag in cpp layer which will ensure that the callback will only be called if session close is initiated by js layer.
2. Added logic to call js layer only if nghttp2_wants_read and nghttp2_wants_write is false

Both of the above things when combined, will ensure that the new callback `onstreamafterwrite_string` will only be called once during the whole lifecycle of a session.

Regarding Failing tests, I am able to reproduce this occasionally i.e once in 10 run but not able to find root cause yet. I will be investigating further about RC.

mcollina reviewed (approved):
lgtm

mcollina commented:
CI doesn't seem happy here, can you take a look?

pandeykushagra51 commented:
> CI doesn't seem happy here, can you take a look?

yeah I am planning to do by today

pandeykushagra51 commented:
### Important Update

## Why Test Updated:

# Original issue with test:
When a client closes an HTTP/2 stream by sending an RST_STREAM frame, the server receives this frame and should trigger the `onStreamClose` callback. In the test scenario, the following sequence occurred:

- The client closed a stream with an error code
- Immediately after, the client closed the entire connection
- This sent a GOAWAY frame to the server
- The server began closing its session

Under the original implementation, the server session closed prematurely before the `onStreamClose` callback could be properly executed for server session. This caused the test to "pass" incorrectly because the expected error was never thrown - the session was terminated and server was closed before the error handling logic could run.

## Simple Way to Verify Test Incorrectness:
Update the test to close server and client with some delay (say 5s), i.e replace line 55,56 of original test of `test/parallel/test-http2-client-rststream-before-connect.js` with below code.
```javascript
setTimeout(() => {
  client.close();
  server.close();
}, 5000); 
```
I have tested with above changes on nodejs version 23.7. and the test failed with this change.

This change exposes the hidden bug because:

- The client sends RST_STREAM to terminate the stream with an error code
- With the delay, the server has plenty of time to process this frame
- The onStreamClose callback fully executes and throws the stream error
- Since there's no error handler, the test fails with an uncaught exception

This confirms that the original test was passing incorrectly by relying on a race condition:

- Original behavior: Client closes connection so quickly that server session terminates before fully processing the RST_STREAM frame
- With delay: Server fully processes all frames and correctly throws the expected error

## How current graceful closure fix worked and exposes test in correctness:
The new implementation improves HTTP/2 session closure by ensuring graceful termination:

- When the server receives a GOAWAY frame from the client, it now checks if there's any pending data in the nghttp2 layer
- If pending operations exist (like an unprocessed RST_STREAM frame), the session remains open until all data is processed
- The session only closes after all callbacks (including `onStreamClose`) have completed execution

With this fix, the `onStreamClose` callback now runs properly, correctly throwing the expected stream error. This reveals that the test was passing incorrectly before because of premature session termination.

So this also fix the Flakiness of current test case.

mcollina reviewed (approved):
lgtm

pimterry commented on src/node_http2.cc:
```diff
@@ -3552,6 +3575,39 @@ void Initialize(Local<Object> target,
+
+  int want_write = nghttp2_session_want_write(session);
+  int want_read = nghttp2_session_want_read(session);
+  bool should_notify = (want_write == 0 && want_read == 0);
```
We currently call this new `CheckAndNotifyJSAfterWrite` method only after writes, but we're checking the read _and_ write states. Presumably there are cases where `want_write == 0 && want_read != 0`, and so this status would only change to allow session shutdown after a read. We don't check this after reads anywhere AFAICT so we'd never handle that.

I'm not totally sure exactly when this would happen, but it seems quite possible, and it would result in the session entering graceful close and never leaving it. For example, I imagine (I haven't tested this!) if a client connects to server, sends request and starts a close - the client is then waiting for data but has nothing to write. The server will respond, but the client will never write anything else, and with this change the client could now get stuck in graceful close indefinitely without every actually closing the connection.

I haven't tested this, so maybe there's some other reason that never happens, but does it sound like there might be plausible routes to this state? It would be very interesting to test this if possible. If that is a real issue, we'd want to add a call to this method in the read logic somewhere too (e.g. after ConsumeHTTP2Data).

pandeykushagra51 replied:
I will investigate further and update here

pandeykushagra51 replied:
Yeah this is important consideration. I will test this behaviour and confirm.

pandeykushagra51 replied:
One of the case I can think of when session send settings page and wait for acknowledgement of settings page. This can be a deadlock condition.
Let me test the behaviour in this case as well.

pimterry replied:
Ok, that makes sense. It's still useful to continue to check this condition regardless, rather than relying on that undocumented internal behaviour in nghttp2. Lets keep the check but just add a comment here to document this, and make it clear that we expect this to always be zero after the goaway write has completed.

pandeykushagra51 replied:
okay fine, I will make add the comment

pandeykushagra51 replied:
added comment

pimterry reviewed (changes requested):
This looks pretty good. I'd like to confirm we don't get stuck in the read scenario I've mentioned here, and there are some small style tweaks I think would be helpful too, but overall the approach looks solid.

Good catch spotting & explaining this issue and nice work on fixing it @pandeykushagra51! This is a tricky bit of behaviour and a tidy solution.

pandeykushagra51 commented on src/node_http2.cc:
```diff
@@ -3552,6 +3575,39 @@ void Initialize(Local<Object> target,
+  Debug(session, "Setting graceful close initiated flag");
+}
+
+bool Http2Session::CheckAndNotifyJSAfterWrite() {
```
This really makes sense, after implementation of optimisation, the function behaviour is changed so it is good update the name.

pandeykushagra51 replied:
done

pandeykushagra51 commented on src/node_http2.cc:
```diff
@@ -3552,6 +3575,39 @@ void Initialize(Local<Object> target,
+  Debug(session, "Setting graceful close initiated flag");
+}
+
+bool Http2Session::CheckAndNotifyJSAfterWrite() {
```
Yeah you are write, initially I created function with some different logic and later on I updated the function behaviour but forgot to update return type.
I will be updating this soon.

pandeykushagra51 replied:
done

pandeykushagra51 commented:
Hey guys could you please check the last commit-queue.

`Add commit-queue-squash label to land the PR as one commit, or commit-queue-rebase to land as separate commits.`

pandeykushagra51 commented:
Thanks a lot @lpinca @jasnell @mcollina @pimterry.
It was really great working on node. Hope to have some more impactful contribution in near future.

pandeykushagra51 commented:
Also please take a look at #57586. 
it is also having commit-queue-failed status but have all checks passed and 3 approval.
